<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="entrain">
<title>Entrain ligand-velocity analysis • entrain</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Entrain ligand-velocity analysis">
<meta property="og:description" content="entrain">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">entrain</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/entrain_traj_manno.html">Entrain ligand-trajectory analysis</a>
    <a class="dropdown-item" href="../articles/entrain_velo.html">Entrain ligand-velocity analysis</a>
    <a class="dropdown-item" href="../articles/faqs.html">FAQ</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Entrain ligand-velocity analysis</h1>
            
      
      
      <div class="d-none name"><code>entrain_velo.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="entrain-ligand-velocity-analysis-from-an-anndata-object-">Entrain ligand-velocity analysis from an Anndata object.<a class="anchor" aria-label="anchor" href="#entrain-ligand-velocity-analysis-from-an-anndata-object-"></a>
</h2>
<p>This document outlines a basic Entrain analysis starting from an
<code>scverse</code> <code>anndata</code> object with pre-calculated
velocities. By the end of this document, you will identify ligands that
are driving the velocities in your data.</p>
<p><strong>Prior assumptions:</strong> Entrain-Velocity Analysis
requires the following:</p>
<ol style="list-style-type: decimal">
<li>Single-Cell RNA data on a dataset of cells differentiating as well
as data on their microenvironmental niche. The niche can be contained in
a separate dataset (e.g., if you have sequenced in a separate capture)
or in the same dataset as the differentiating cells.</li>
<li>A <code>.h5ad</code> object produced by scvelo that contains robust
velocity data. Please ensure that your biology you are interested in is
conducive to <a href="https://doi.org/10.15252/msb.202110282" class="external-link">producing
trustworthy velocities.</a>.</li>
<li>You have (some) idea of which cell clusters comprise the ‘niche’, or
ligand-expressing cells, in your dataset.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Seurat"</span>); <span class="fu">library</span>(<span class="st">"SeuratWrappers"</span>);<span class="fu">library</span>(<span class="st">"reticulate"</span>); <span class="fu">library</span>(<span class="st">"dplyr"</span>), <span class="fu">library</span>(<span class="st">"circlize"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"entrain"</span>)</span></code></pre></div>
<div class="section level3">
<h3 id="pre-processing-and-loading-required-data-">Pre-processing and loading required data.<a class="anchor" aria-label="anchor" href="#pre-processing-and-loading-required-data-"></a>
</h3>
<div class="section level4">
<h4 id="conda-setup-">Conda setup.<a class="anchor" aria-label="anchor" href="#conda-setup-"></a>
</h4>
<p>Because RNA velocity is natively done in python, this workflow
involves calling python scripts from R to process the anndata object.
The interfacing between python and R is handled by the package
<code>reticulate</code>. We recommend using conda to setup a python
environment that contains the required packages for velocity
analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reticulate::conda_create("r-reticulate-sc")</span></span>
<span><span class="co"># reticulate::conda_install("r-reticulate-sc", "scanpy", "leidenalg", "python-igraph", "adjustText", channel="conda-forge")</span></span>
<span><span class="co"># reticulate::conda_install("r-reticulate-sc", "scvelo", channel="bioconda") #scvelo not on default conda.</span></span>
<span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_condaenv</a></span><span class="op">(</span><span class="st">"r-reticulate-sc"</span><span class="op">)</span>  <span class="co"># points to a conda installation that contains scanpy, scvelo, and anndata.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="loading-in-required-data">Loading in required data<a class="anchor" aria-label="anchor" href="#loading-in-required-data"></a>
</h4>
<p>We will download a developing mouse brain atlas dataset at gestation
day 10/11. This dataset contains both RNA expression and RNA velocities
needed for our analysis. The following code block downloads the .gz
file, decompresses it, and converts it into a Seurat object for
analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://mojaveazure.github.io/seurat-disk/" class="external-link">"SeuratDisk"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">3600</span><span class="op">)</span></span>
<span><span class="va">h5ad_path</span> <span class="op">&lt;-</span> <span class="st">"Manno_E10_E11.h5ad"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/Manno_E10_E11.h5ad.gz"</span>, <span class="st">"Manno_E10_E11.h5ad.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"gunzip Manno_E10_E11.h5ad.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/Convert.html" class="external-link">Convert</a></span><span class="op">(</span><span class="st">"manno_entrain.h5ad"</span>, dest <span class="op">=</span> <span class="st">"h5seurat"</span>, overwrite <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/LoadH5Seurat.html" class="external-link">LoadH5Seurat</a></span><span class="op">(</span><span class="st">"Manno_E10_E11.h5seurat"</span>, assays<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<p>Entrain relies on the NicheNet database of ligand-receptor-gene
networks for the prior knowledge needed to make conclusions about
environmentally-influenced trajectories. The genes here have been
pre-converted from the original human genes to mouse orthologs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lr_network_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/lr_network_mm.rds"</span><span class="op">)</span><span class="op">)</span> <span class="co">#</span></span>
<span><span class="va">ligand_target_matrix_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/ligand_target_matrix_mm.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="data-at-a-glance">Data at a glance<a class="anchor" aria-label="anchor" href="#data-at-a-glance"></a>
</h4>
<p>The data consists of cells in the developing mouse brain at day 10
after gestation. This comprises a population of neuroblasts rapidly
differentiating to neurons (our cells we are going to analyse), as well
as their complex microenvironment made up of cells from the endoderm,
mesoderm, fibroblastic, blood, and immune compartments.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_clusters</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">Class</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DiscretePalette.html" class="external-link">DiscretePalette</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n_clusters</span>, palette <span class="op">=</span> <span class="st">"alphabet2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"Class"</span>, cols<span class="op">=</span><span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="entrain_velo_files/figure-html/Dataset%20at%20a%20glance-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="run-entrain-from-a-velocity-anndata-object-">Run Entrain from a velocity anndata object.<a class="anchor" aria-label="anchor" href="#run-entrain-from-a-velocity-anndata-object-"></a>
</h3>
<p>Entrain fully integrates the scvelo package for velocity analysis.
Below are the main steps:</p>
<div class="section level4">
<h4 id="clustering-velocities">Clustering Velocities<a class="anchor" aria-label="anchor" href="#clustering-velocities"></a>
</h4>
<ol style="list-style-type: decimal">
<li>First, cluster velocities into groups of similar velocity dynamics.
The resultant velocity clusters (or <code>vclusters</code>) are a
separate entity from the usual cell annotation clusters. Cells of the
same celltype cluster may not necessarily be in the same velocity
cluster.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entrain_cluster_velocities.html">entrain_cluster_velocities</a></span><span class="op">(</span><span class="va">h5ad_path</span>, resolution<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/entrain_velocity_clusters.png" width="692"></p>
<p>We can see that the velocity clusters roughly, but not exactly,
correspond to cell type clusters, indicating that velocities are
moderately correlated with cell type.</p>
</div>
<div class="section level4">
<h4 id="fitting-ligands-to-velocities">Fitting Ligands to Velocities<a class="anchor" aria-label="anchor" href="#fitting-ligands-to-velocities"></a>
</h4>
<p>The fitting is encapsulated in the single function
<code><a href="../reference/entrain_velocity.html">entrain_velocity()</a></code>, shown below. This function firstly runs
<code>scvelo.tl.recover_dynamics()</code> to rank genes based on their
fit-likelihoods, a measure of a gene’s fit to the ordinary differential
equation models that underlie RNA velocity derivation. Higher values
denote more confidence in the velocity. Next, it runs Entrain analysis
on the velocity likelihoods on a cluster-by-cluster basis.</p>
<p>We will also save the resulting .h5ad object for plotting our results
afterwards.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sender_cluster_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Blood"</span>, <span class="st">"Fibroblast"</span>, <span class="st">"Immune"</span>, <span class="st">"Mesoderm"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Class"</span></span>
<span><span class="va">save_adata</span> <span class="op">=</span>  <span class="st">"../../entrain_vignette_data/manno_entrain_result.h5ad"</span></span>
<span><span class="va">obj_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/entrain_velocity.html">entrain_velocity</a></span><span class="op">(</span>receiver_obj <span class="op">=</span> <span class="va">obj</span>,</span>
<span>                          adata <span class="op">=</span> <span class="va">adata</span>,</span>
<span>                          save_adata <span class="op">=</span> <span class="va">save_adata</span>,</span>
<span>                          sender_cluster_key <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>                          sender_clusters <span class="op">=</span> <span class="va">sender_cluster_names</span>,</span>
<span>                          lr_network <span class="op">=</span> <span class="va">lr_network_mm</span>,</span>
<span>                          ligand_target_matrix <span class="op">=</span> <span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualizing-results">Visualizing results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h3>
<div class="section level4">
<h4 id="at-a-glance">At a glance<a class="anchor" aria-label="anchor" href="#at-a-glance"></a>
</h4>
<p>We can visualize the top ranked ligands at a glance with the function
<code><a href="../reference/plot_velocity_ligands.html">plot_velocity_ligands()</a></code>. By default, the function only
visualizes results with positive variance explained, with the assumption
that negative variance explained denotes poor model accuracy.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_velocity_ligands.html">plot_velocity_ligands</a></span><span class="op">(</span>adata<span class="op">=</span><span class="va">save_adata</span>,</span>
<span>                      plot_output_path <span class="op">=</span> <span class="st">"manno_entrain.png"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/manno_entrain.png" width="1500"> This suggests a
number of ligands responsible for neurogenic differentiation processes.
These are well-established in literature as key drivers for cell
differentiation.</p>
</div>
<div class="section level4">
<h4 id="in-detail">In detail<a class="anchor" aria-label="anchor" href="#in-detail"></a>
</h4>
<p>Entrain-Velocity provides detailed plotting functions for visualizing
extracellular and intracellular regulation predicted to be driving the
observed velocities.</p>
<div class="section level5">
<h5 id="sender-cell-visualization">Sender Cell Visualization<a class="anchor" aria-label="anchor" href="#sender-cell-visualization"></a>
</h5>
<p>Visualization of cells that express the ligands found in our
analysis. Cells are coloured based on mean expression of the
<code>top_n_ligands</code> predicted to be driving velocities in the
velocity cluster.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vclusters_to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vcluster0"</span>, <span class="st">"vcluster1"</span>, <span class="st">"vcluster2"</span>, <span class="st">"vcluster3"</span><span class="op">)</span></span>
<span><span class="va">save_adata</span> <span class="op">=</span>  <span class="st">"../../entrain_vignette_data/manno_entrain.h5ad"</span></span>
<span><span class="fu"><a href="../reference/plot_sender_influence_velocity.html">plot_sender_influence_velocity</a></span><span class="op">(</span>adata<span class="op">=</span> <span class="va">save_adata</span>,</span>
<span>                                   velocity_clusters <span class="op">=</span> <span class="va">vclusters_to_plot</span>,</span>
<span>                                   title <span class="op">=</span> <span class="st">"velocity_sender_influences.png"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/entrain_velocity_influence.png" width="4800">
#### Visualizing ligand importances Visualization of ligands and their
relative importances, or contribution, towards each
<code>vcluster</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/velocity_ligand_importance_heatmap.html">velocity_ligand_importance_heatmap</a></span><span class="op">(</span><span class="va">obj_v</span>,</span>
<span>                                     n_top_ligands<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="entrain_velo_files/figure-html/Importances%20heatmap-1.png" width="960"></p>
</div>
<div class="section level5">
<h5 id="visualizing-high-velocity-likelihood-genes">Visualizing high velocity likelihood genes<a class="anchor" aria-label="anchor" href="#visualizing-high-velocity-likelihood-genes"></a>
</h5>
<p>Commonly, a small proportion of the total genes contribute the
majority of the observed RNA velocity dynamics. You may wish to
visualize these genes by plotting the top likelihood genes. These genes
are the ones found by <code>scvelo</code> to most robustly fit the
underlying theoretical model. The function
<code><a href="../reference/plot_velocity_likelihoods_heatmap.html">plot_velocity_likelihoods_heatmap()</a></code> plots, as a heatmap,
the <code>n_top_genes</code> ranked by velocity likelihood for each
velocity cluster.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/plot_velocity_likelihoods_heatmap.html">plot_velocity_likelihoods_heatmap</a></span><span class="op">(</span><span class="va">obj_v</span>,</span>
<span>                                     n_top_genes<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="entrain_velo_files/figure-html/Likelihood%20heatmap-1.png" width="960"></p>
<p>A good sanity check is to confirm that the top genes are what we
would expect from the biological system. In this case, FEDSFSDFSD</p>
</div>
<div class="section level5">
<h5 id="visualizing-intracellular-regulation-downstream-of-the-ligand-receptor-pair-">Visualizing intracellular regulation downstream of the
ligand-receptor pair.<a class="anchor" aria-label="anchor" href="#visualizing-intracellular-regulation-downstream-of-the-ligand-receptor-pair-"></a>
</h5>
<p>The function <code>plot_velocity_ligand_targets</code> extracts
regulatory information from <code>NicheNetR</code> to visualize the
predicted regulatory linkages between the <code>n_top_ligands</code> and
the <code>n_top_targets</code> high velocity likelihood genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/plot_velocity_ligand_targets.html">plot_velocity_ligand_targets</a></span><span class="op">(</span><span class="va">obj_v</span>, vcluster<span class="op">=</span><span class="st">"vcluster1"</span>, n_top_ligands<span class="op">=</span><span class="fl">10L</span>, ligand_target_matrix<span class="op">=</span><span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span>
<span><span class="va">g</span></span></code></pre></div>
<p><img src="figures/velocity_targets.png" width="1500"></p>
</div>
<div class="section level5">
<h5 id="circos-plot">Circos plot<a class="anchor" aria-label="anchor" href="#circos-plot"></a>
</h5>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weighted_networks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/weighted_networks_mm.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/draw_entrain_circos.html">draw_entrain_circos</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">obj_v</span>,</span>
<span>                    vcluster<span class="op">=</span> <span class="st">"vcluster3"</span>,</span>
<span>                    lr_network <span class="op">=</span> <span class="va">lr_network_mm</span>,</span>
<span>                    weighted_networks <span class="op">=</span> <span class="va">weighted_networks</span><span class="op">)</span></span></code></pre></div>
<p><img src="entrain_velo_files/figure-html/show%20plot-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Wunna Kyaw.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
