<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="entrain">
<title>Entrain ligand-trajectory analysis • entrain</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Entrain ligand-trajectory analysis">
<meta property="og:description" content="entrain">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">entrain</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/entrain_traj_manno.html">Entrain ligand-trajectory analysis</a>
    <a class="dropdown-item" href="../articles/entrain_velo.html">Entrain ligand-velocity analysis</a>
    <a class="dropdown-item" href="../articles/faqs.html">FAQ</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Entrain ligand-trajectory analysis</h1>
            
      
      
      <div class="d-none name"><code>entrain_traj_manno.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="ligand-trajectory-analysis-from-a-seurat-object-">Ligand-Trajectory Analysis from a Seurat object.<a class="anchor" aria-label="anchor" href="#ligand-trajectory-analysis-from-a-seurat-object-"></a>
</h2>
<p>This document outlines a basic Entrain analysis starting from a
Seurat object. By the end of this document, you will identify ligands
that are driving the trajectories in your data.</p>
<p><strong>Prior assumptions:</strong> Entrain-Trajectory Analysis
requires the following:</p>
<ol style="list-style-type: decimal">
<li>You have Single-Cell RNA data on a dataset of cells differentiating
as well as data on their microenvironmental niche. The niche can be
contained in a separate dataset (e.g., if you have sequenced in a
separate capture) or in the same dataset as the differentiating
cells.</li>
<li>You have an idea of which cell clusters comprise the ‘niche’, or
ligand-expressing cells, in your dataset.</li>
<li>Your dataset has sufficiently sampled cells from all beginning,
intermediate and terminal stages in the differentiation continuum. If
not, your analysis will be spurious during the trajectory learning
step.</li>
</ol>
</div>
<div class="section level2">
<h2 id="pre-processing-and-loading-required-data-">Pre-processing and loading required data.<a class="anchor" aria-label="anchor" href="#pre-processing-and-loading-required-data-"></a>
</h2>
<div class="section level3">
<h3 id="loading-in-libraries">Loading in libraries<a class="anchor" aria-label="anchor" href="#loading-in-libraries"></a>
</h3>
<p>If you haven’t installed <code>Entrain</code> yet, see the home
page.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://mojaveazure.github.io/seurat-disk/" class="external-link">"SeuratDisk"</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SeuratWrappers"</span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span>;</span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"entrain"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-in-required-data">Loading in required data<a class="anchor" aria-label="anchor" href="#loading-in-required-data"></a>
</h3>
<p>We will download a developing mouse brain atlas dataset at gestation
day 10/11. The following code block downloads the .gz file to the
working directory, decompresses it, and converts it into a Seurat object
for analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">3600</span><span class="op">)</span></span>
<span><span class="co">#download.file("https://zenodo.org/record/7233173/files/Manno_E10_E11.rds", "Manno_E10_E11.rds")</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Manno_E10_E11.rds"</span><span class="op">)</span></span></code></pre></div>
<p>Entrain relies on the NicheNet database of ligand-receptor-gene
networks for the prior knowledge needed to make conclusions about
environmentally-influenced trajectories. The genes here have been
pre-converted from the original human genes to mouse orthologs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lr_network_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/lr_network_mm.rds"</span><span class="op">)</span><span class="op">)</span> <span class="co">#</span></span>
<span><span class="va">ligand_target_matrix_mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/7233173/files/ligand_target_matrix_mm.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-at-a-glance">Data at a glance<a class="anchor" aria-label="anchor" href="#data-at-a-glance"></a>
</h3>
<p>The data consists of cells in the developing mouse brain at day 10
after gestation. This comprises a population of neuroblasts rapidly
differentiating to neurons (our cells we are going to analyse), as well
as their complex microenvironment made up of cells from the endoderm,
mesoderm, fibroblastic, blood, and immune compartments.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_clusters</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">Class</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DiscretePalette.html" class="external-link">DiscretePalette</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n_clusters</span>, palette <span class="op">=</span> <span class="st">"alphabet2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">obj</span>, group.by <span class="op">=</span> <span class="st">"Class"</span>, cols<span class="op">=</span><span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="entrain_traj_manno_files/figure-html/Dataset%20at%20a%20glance-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="run-entrain">Run Entrain<a class="anchor" aria-label="anchor" href="#run-entrain"></a>
</h2>
<div class="section level3">
<h3 id="entrain-from-a-fresh-seurat-object-">Entrain from a fresh Seurat object.<a class="anchor" aria-label="anchor" href="#entrain-from-a-fresh-seurat-object-"></a>
</h3>
<p>Entrain fully integrates the Monocle3 package to learn trajectories.
All steps of Monocle, as well as Entrain analysis, are encapsulated in
an interactive workflow with the function
<code><a href="../reference/get_traj_ligands_monocle.html">get_traj_ligands_monocle()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj_entr</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get_traj_ligands_monocle.html">get_traj_ligands_monocle</a></span><span class="op">(</span><span class="va">obj</span>,</span>
<span>                                   sender_cluster_key <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>                                   lr_network <span class="op">=</span> <span class="va">lr_network_mm</span>, ligand_target_matrix<span class="op">=</span><span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span></code></pre></div>
<p>This function launches three steps:</p>
<ol style="list-style-type: decimal">
<li><p>Call Monocle3 to learn trajectories.</p></li>
<li><p>Second, select the cell clusters corresponding to the
microenvironment.</p></li>
</ol>
<details><summary><strong>Click to show interactive example:</strong>
</summary><style>
.shiny-app-frame {
  position: relative;
  height: 800px;
  width: 950;
  margin: auto;
}
.shiny-app-frame iframe {
  width: 100%;
  height: 100%;
  border: none;
}
</style>
<div class="shiny-app-frame">
<iframe src="https://wunna.shinyapps.io/sender_selection_shinyapp/">
</iframe>
</div>
</details><ol start="3" style="list-style-type: decimal">
<li>Third, select the trajectory branches that you wish to analyze with
Entrain.</li>
</ol>
<details><summary><strong>Click to show interactive example:</strong>
</summary><p>In this example, we can see that Monocle has captured some spurious
trajectories (e.g., excessive branching in the radial glial cells) as
well as some potentially biological ones (e.g., neuroblast and
neurons).</p>
Entrain allows us to analyze only the trajectories that we are confident
in and discard the rest; in this case we are more confident that the
neuroblast differentiation trajectory is real. This trajectory is
described by the cells located in the obvious line segment in the left
of the plot below. You can simulate selecting that trajectory in the
interactive window.
<div class="shiny-app-frame">
<iframe src="https://wunna.shinyapps.io/deploy/">
</iframe>
</div>
</details>
</div>
<div class="section level3">
<h3 id="run-entrain-on-a-pre-generated-monocle-trajectory">Run Entrain on a pre-generated Monocle trajectory<a class="anchor" aria-label="anchor" href="#run-entrain-on-a-pre-generated-monocle-trajectory"></a>
</h3>
<p>If you have already run trajectory analysis on your dataset, you can
insert the existing trajectory into <code>Entrain</code>.</p>
<details><summary><strong>Click to show</strong>
</summary><p>If you already have a pre-generated trajectory generated with the
<code>monocle3</code> package, Entrain allows you to start from that
dataset instead of re-running the process. The pre-generated trajectory
should have <code><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">monocle3::learn_graph()</a></code> and
<code><a href="https://rdrr.io/pkg/monocle3/man/order_cells.html" class="external-link">monocle3::order_cells()</a></code> called on it already, and should be
passed into the argument
<code>get_traj_ligands_monocle(..., cds = monocle_cds)</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj_entr</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/get_traj_ligands_monocle.html">get_traj_ligands_monocle</a></span><span class="op">(</span><span class="va">obj</span>, cds <span class="op">=</span> <span class="va">monocle_cds</span>,</span>
<span>                                   sender_cluster_key <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>                                   lr_network <span class="op">=</span> <span class="va">lr_network_mm</span>, ligand_target_matrix<span class="op">=</span><span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="run-entrain-in-a-script-">Run Entrain in a script.<a class="anchor" aria-label="anchor" href="#run-entrain-in-a-script-"></a>
</h3>
This is most useful when you are running on a compute cluster without a
GUI to interact with.
<details><summary><strong>Click to show</strong>
</summary><p>First, we will run a standard monocle workflow to learn the
trajectory.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">monocle_cds</span> <span class="op">&lt;-</span> <span class="fu">SeuratWrappers</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratWrappers/man/as.cell_data_set.html" class="external-link">as.cell_data_set</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span>
<span><span class="va">monocle_cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html" class="external-link">preprocess_cds</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">monocle_cds</span>, num_dim <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">monocle_cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html" class="external-link">reduce_dimension</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">monocle_cds</span><span class="op">)</span></span>
<span><span class="va">monocle_cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/cluster_cells.html" class="external-link">cluster_cells</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">monocle_cds</span>, reduction_method <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span>
<span><span class="va">monocle_cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">learn_graph</a></span><span class="op">(</span><span class="va">monocle_cds</span>,</span>
<span>                                     use_partition <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                     close_loop <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                     learn_graph_control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                         minimal_branch_len <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                         ncenter <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Since this is run in a script, we cannot view the trajectory
interactively. Instead, the below code saves the trajectory and node
labels to <code>.png</code>. You can then download the <code>.png</code>
from the cluster and open it on your local machine. This allows you to
validate that the trajectory makes biological sense before running
Entrain analysis, and if so, shows you the node labels that you will
need to run Entrain without a GUI.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>node_coords <span class="ot">&lt;-</span> monocle_cds<span class="sc">@</span>principal_graph_aux<span class="sc">$</span>UMAP<span class="sc">$</span>dp_mst <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">node =</span> <span class="fu">rownames</span>(.)) <span class="sc">%&gt;%</span> stats<span class="sc">::</span><span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"node"</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>fig<span class="ot">&lt;-</span>monocle3<span class="sc">::</span><span class="fu">plot_cells</span>(monocle_cds, <span class="at">color_cells_by=</span><span class="st">"Class"</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">data =</span> node_coords, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y)) <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    ggrepel<span class="sc">::</span><span class="fu">geom_text_repel</span>(<span class="at">data =</span> node_coords, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">label=</span>node)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">size=</span><span class="fl">2.5</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(fig, <span class="st">"monocle_plot.png"</span>)</span></code></pre></div>
<p><img src="entrain_traj_manno_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>As before, we can see that Monocle has captured some spurious
trajectories (e.g., excessive branching in the radial glial cells) as
well as some potentially biological ones (e.g., neuroblast and
neurons).</p>
<p>The neuroblast trajectory we are confident in is located between
nodes <code>Y_5</code> and <code>Y_27</code>.</p>
<p>We have already highlighted nodes <code>Y_5</code> and
<code>Y_27</code> in the plot above, but when running on your data you
will have to decide which nodes denote the beginning and terminus of
your trajectory of interest.</p>
<p>We will also define the radial glial cells at node <code>Y_7</code>
as the root cells, denoting the progenitor cell types. We can then run
Entrain programmatically by setting <code>root_pr_nodes = "Y_7"</code>
and <code>path_nodes = c("Y_5", "Y_27")</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj_entr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_traj_ligands_monocle.html">get_traj_ligands_monocle</a></span><span class="op">(</span><span class="va">obj</span>, cds <span class="op">=</span> <span class="va">monocle_cds</span>,</span>
<span>                                   sender_cluster_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Blood"</span>, <span class="st">"Endoderm"</span>, <span class="st">"Fibroblast"</span>,</span>
<span>                                                              <span class="st">"Immune"</span>, <span class="st">"Mesoderm"</span><span class="op">)</span>,</span>
<span>                                   sender_cluster_key <span class="op">=</span> <span class="st">"Class"</span>,</span>
<span>                                   root_pr_nodes <span class="op">=</span> <span class="st">"Y_7"</span>,</span>
<span>                                   path_nodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Y_5"</span>, <span class="st">"Y_27"</span><span class="op">)</span>,</span>
<span>                                   lr_network<span class="op">=</span><span class="va">lr_network_mm</span>, ligand_target_matrix<span class="op">=</span><span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-results">Visualizing results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h2>
<div class="section level3">
<h3 id="at-a-glance">At a glance<a class="anchor" aria-label="anchor" href="#at-a-glance"></a>
</h3>
<p>We can visualize the top ranked ligands at a glance with the
following function, replacing “celltype” with the relevant column name
in <code>obj@meta.data</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_ligand_trajectories.html">plot_ligand_trajectories</a></span><span class="op">(</span><span class="va">obj_entr</span>, color_cells_by <span class="op">=</span> <span class="st">"pseudotime"</span>, group_label_size<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Cells aren't colored in a way that allows them to be grouped.</span></span></code></pre>
<p><img src="entrain_traj_manno_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This suggests a number of ligands responsible for neuroblast
differentiation. These are well-established in literature as key drivers
for neurogenesis. These ligands are relatively low in expression, and so
would be difficult to pick up with a standard ligand-receptor analysis
compared to more strongly expressed, but less biologically impactful
ligands. For example, NicheNet detects <code>Bmp5</code> in it’s top 10
results, but is unable to pick up
<code>Bdnf, Slit2, Dll1, and Tgfb2</code> - all of which are
microenvironmental signals supported in neurobiology literature.</p>
</div>
<div class="section level3">
<h3 id="in-detail">In detail<a class="anchor" aria-label="anchor" href="#in-detail"></a>
</h3>
<p>We can visualize the invididual ligand-gene relationships that
contributing most to the trajectory dynamics. This data is extracted
from the underlying NicheNet database that Entrain uses for fitting.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">paths</span> <span class="op">&lt;-</span> <span class="va">obj_entr</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">entrain</span><span class="op">$</span><span class="va">paths</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">plot_lig</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/plot_ligand_targets.html">plot_ligand_targets</a></span><span class="op">(</span><span class="va">obj_entr</span>, path<span class="op">=</span><span class="va">paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, ligand_target_matrix <span class="op">=</span> <span class="va">ligand_target_matrix_mm</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the top covarying genes of a trajectory, including
genes that are not influenced by extracellular signaling.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_genes</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/plot_covarying_genes_scatter.html">plot_covarying_genes_scatter</a></span><span class="op">(</span><span class="va">obj_entr</span>, path<span class="op">=</span><span class="va">paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, color_key<span class="op">=</span><span class="st">"Class"</span>, n_top_genes <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">plot_genes</span></span></code></pre></div>
<p><img src="entrain_traj_manno_files/figure-html/unnamed-chunk-9-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="cell-wise-influences">Cell-wise influences<a class="anchor" aria-label="anchor" href="#cell-wise-influences"></a>
</h3>
<p>Trajectory branches represent a continuum of states, which may
comprise varying degrees of environmental dependence along the
continuum. We can visualize where the environmental influence is
occurring in this continuum to generate hypotheses about cell states
which more or less prone to environmental influence.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj_entr</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/cellwise_influences.html">cellwise_influences</a></span><span class="op">(</span><span class="va">obj_entr</span>,</span>
<span>                              ligand_target_matrix<span class="op">=</span><span class="va">ligand_target_matrix_mm</span>,</span>
<span>                              step_size<span class="op">=</span><span class="fl">0.10</span>,</span>
<span>                              window_pct<span class="op">=</span><span class="fl">0.30</span>,</span>
<span>                              n_top_ligands<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_ligand_influences.html">plot_ligand_influences</a></span><span class="op">(</span><span class="va">obj_entr</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/manno_entrain_influences.png" width="400px" height="400px" style="display: block; margin: auto;"></p>
<p>Our influence analysis highlights the areas most strongly influenced
by the environment in dark red. These regions coincide with the border
of neuroblast and neuron cell type label, and on the border of radial
glia and neuroblast cell type labels. This is despite our analysis being
agnostic to labelling.</p>
<!-- https://github.com/r-lib/pkgdown/issues/838 -->
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Wunna Kyaw.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
